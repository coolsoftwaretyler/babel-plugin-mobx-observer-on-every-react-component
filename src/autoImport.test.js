import { expect, test, describe } from "bun:test";
import { transform } from "@babel/core";
import pluginSyntaxJsx from "@babel/plugin-syntax-jsx";
import pluginTransReactJsx from "@babel/plugin-transform-react-jsx";
import autoObserverPlugin from "./index.ts";

const runTransform = (input) => {
  return transform(input, {
    plugins: [pluginSyntaxJsx, pluginTransReactJsx, autoObserverPlugin],
  });
};

describe("Mobx Observer Babel Plugin", () => {
  describe("observer imports", () => {
    describe("when there is no observer import", () => {
      describe("and no react components", () => {
        test("it does nothing", () => {
          const source = `function MyFunctionDeclaration() { return 0 }`;
          const out = runTransform(source);

          expect(out.code).toMatchSnapshot();
        });
      });
      describe("and there are react components", () => {
        test("it adds an observer import", () => {
          const source = `function MyComponentDeclaration() { return <div>Hello World</div>; }`;
          const out = runTransform(source);

          expect(out.code).toMatchSnapshot();
        });
      });
      describe('and we have other imports from mobx-react', () => {
        test('it adds the observer import', () => {
          const source = `import { inject } from "mobx-react"; function MyComponentDeclaration() { return <div>Hello World</div>; }`;
          const out = runTransform(source);

          expect(out.code).toMatchSnapshot();
        })
      })
    });
    describe("when there is already an observer import", () => {
      describe("and no react components", () => {
        test("it does nothing", () => {
          const source = `import { observer } from "mobx-react"; function MyFunctionDeclaration() { return 0 }`;
          const out = runTransform(source);

          expect(out.code).toMatchSnapshot();
        });
      });
      describe("and react components but unwrapped", () => {
        test("it does not add the import, but wraps the component", () => {
          const source = `import { observer } from "mobx-react"; function MyComponentDeclaration() { return <div>Hello World</div>; }`;
          const out = runTransform(source);

          expect(out.code).toMatchSnapshot();
        });
      });
      describe("and react components wrapped", () => {
        test("it takes no action", () => {
            const source = `import { observer } from "mobx-react"; function MyComponentDeclaration() { return <div>Hello World</div>; }`;
            const out = runTransform(source);
    
            expect(out.code).toMatchSnapshot();
        })
      })
      describe('and we have other imports from mobx-react', () => {
        test('it does not add the observer import', () => {
          const source = `import { observer, inject } from "mobx-react"; function MyComponentDeclaration() { return <div>Hello World</div>; }`;
          const out = runTransform(source);

          expect(out.code).toMatchSnapshot();
        })
      })
    });
  });
});
