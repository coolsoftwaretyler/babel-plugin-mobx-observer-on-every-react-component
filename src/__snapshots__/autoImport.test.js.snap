// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Mobx Observer Babel Plugin observer imports when there is no observer import and no react components it does nothing 1`] = `
"function MyFunctionDeclaration() {
  return 0;
}"
`;

exports[`Mobx Observer Babel Plugin observer imports when there is no observer import and there are react components it adds an observer import 1`] = `
"import { observer } from "mobx-react";
function MyComponentDeclaration() {
  return observer(function MyComponentDeclaration() {
    return /*#__PURE__*/React.createElement("div", null, "Hello World");
  });
}"
`;

exports[`Mobx Observer Babel Plugin observer imports when there is already an observer import and no react components it does nothing 1`] = `
"import { observer } from "mobx-react";
function MyFunctionDeclaration() {
  return 0;
}"
`;

exports[`Mobx Observer Babel Plugin observer imports when there is already an observer import and react components but unwrapped it does not add the import, but wraps the component 1`] = `
"import { observer } from "mobx-react";
function MyComponentDeclaration() {
  return observer(function MyComponentDeclaration() {
    return /*#__PURE__*/React.createElement("div", null, "Hello World");
  });
}"
`;

exports[`Mobx Observer Babel Plugin observer imports when there is already an observer import and react components wrapped it takes no action 1`] = `
"import { observer } from "mobx-react";
function MyComponentDeclaration() {
  return observer(function MyComponentDeclaration() {
    return /*#__PURE__*/React.createElement("div", null, "Hello World");
  });
}"
`;
